### PR TITLE
Fixed Grouped Choices for Radios and Checkboxes

### LINKED ISSUE #690: Support Grouped Choice Sets in CheckboxSelectMultiple
* Package version: 1.6.1
* Django version: 1.9 (/all)
* Python version: 2.7 (/all)

### Description:

After Django 1.7, the choice widgets all support nested choices (which render as OPTGROUP tags in the HTML for select widgets, and as nested lists for the checkboxselect widgets). See [this django issue](https://code.djangoproject.com/ticket/20931) for more discussion. In crispy_forms, the default multiple select option works as expected, but the `CheckBoxSelectMultiple` option breaks, causing the choice to be rendered as an escaped python `__repr__`.

### Example Django Crispy Forms code
(Used to generate the screenshot above)
```
from django import forms
from django.forms.widgets import CheckboxChoiceInput
from crispy_forms.layout import *
from crispy_forms.helper import *
from django.template import Template, Context

class MyForm(forms.Form):
    checkboxselectmultiple = forms.MultipleChoiceField(
        widget=forms.CheckboxSelectMultiple,
        choices=(
            ('group1', (
                ('choice1', 'choice1'),
                ('choice2', 'choice2'),
            )),
            ('group2', (
                ('choice1', 'choice1'),
                ('choice2', 'choice2'),
            )),
        )
    )
    multiple = forms.MultipleChoiceField(
        choices=(
            ('group1', (
                ('choice1', 'choice1'),
                ('choice2', 'choice2'),
            )),
            ('group2', (
                ('choice1', 'choice1'),
                ('choice2', 'choice2'),
            )),
        )
    )

    def __init__(self, *args, **kwargs):
        super(MyForm, self).__init__(*args, **kwargs)
        self.helper = FormHelper()
        self.helper.layout = Layout(
            Field('checkboxselectmultiple'),
            Field('multiple'),
        )



template_string = '''\
{% load crispy_forms_tags %}
<html>
    <head>
        <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
        <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" integrity="sha384-rHyoN1iRsVXV4nD0JutlnGaslCJuC7uwjduW9SVrLvRYooPp2bWYgmgJQIXwl/Sp" crossorigin="anonymous">
        <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>
    </head>
    <body>
        <h1>Crispy</h1>
        {% crispy form 'bootstrap3' %}
        <h1>Default</h1>
        {{ form }}
    </body>
</html>
'''

with open('/tmp/foo.html', 'w') as f:
    f.write(Template(template_string).render(context=Context({'form': MyForm()})))
import os; os.system('open /tmp/foo.html')
```

### PR DESCRIPTION
**Pull request by smithdc1**
So here we are 🎉 This fixes the long standing bug with grouped choices sets.

Fixes #690 and Follows #1119 and #813

I'll leave this here for comments, and look at pulling some of the related changes into separate PRs. There's also a few things which I want to take a closer look at. 

- [x] Show image of before / after
- [x] Release notes
- [x] What's happening with the new `required` attribute on `test_use_custom_control_is_used_in_radio_true.html`
- [x] of course the tests fail on Django 2.2

### CODE REVIEW

File: crispy_forms/tests/test_layout_objects.py [hunk 1]
```diff
@@ -517,7 +521,7 @@ def test_multiplecheckboxes(self, settings):
             elif settings.CRISPY_TEMPLATE_PACK == "bootstrap4":
                 assert html.count('custom-control-inline"') == 3
 
-    def test_multiple_checkboxes_unique_ids(self):
+    def test_multiple_checkboxes_unique_ids(self, settings):
```
Comment (smithdc1) [inline crispy_forms/tests/test_layout_objects.py:524]: Can revert this change

File: crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_checkboxes_false.html [hunk 2]
```diff
@@ -1,8 +1,8 @@
 <form method="post">
-    <div class="form-group" id="div_id_checkboxes"><label class=" requiredField" for="">Checkboxes<span
+    <div class="form-group" id="div_id_checkboxes"><label class=" requiredField">Checkboxes<span
                 class="asteriskField">*</span></label>
         <div>
-            <div class="form-check"><input checked="checked" class="form-check-input" id="id_checkboxes_0" name="checkboxes"
```
Comment (smithdc1) [inline crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_checkboxes_false.html:5]: Could be worth pulling this into a separate PR for all template packs.

File: crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_checkboxes_false.html [hunk 3]
```diff
@@ -1,8 +1,8 @@
 <form method="post">
-    <div class="form-group" id="div_id_checkboxes"><label class=" requiredField" for="">Checkboxes<span
```
Comment (smithdc1) [inline crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_checkboxes_false.html:2]: Could pull this change into a separate PR. Is it right that this is blank? 🤔

File: crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_radio_false.html [hunk 4]
```diff
@@ -3,11 +3,11 @@
             select<span class="asteriskField">*</span></label>
         <div>
             <div class="form-check"><input class="form-check-input" id="id_radio_select_0" name="radio_select"
-                    type="radio" value="1"><label class="form-check-label" for="id_radio_select_0">1</label></div>
+                    type="radio" value="1" required><label class="form-check-label" for="id_radio_select_0">1</label></div>
```
Comment (smithdc1) [inline crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_radio_false.html:6]: Is this right?

**Review by smithdc1 — COMMENTED** [review]
A few notes to look at. The pre-work here is proving useful as we can clearly see what the changes are with quite a significant change to the underlying template.

File: crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_radio_false.html (hunk 4 above)
Comment (smithdc1) [reply crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_radio_false.html:6]: Django's own rendering shows renders required here. 

``` html
    <th><label for="id_radio_select_0">Radio select:</label></th>
    <td>
        <ul id="id_radio_select">
            <li><label for="id_radio_select_0"><input type="radio" name="radio_select" value="1" required
                        id="id_radio_select_0"> 1</label></li>
            <li><label for="id_radio_select_1"><input type="radio" name="radio_select" value="2" required
                        id="id_radio_select_1"> 2</label></li>
            <li><label for="id_radio_select_2"><input type="radio" name="radio_select" value="1000" required
                        id="id_radio_select_2"> 1000</label></li>
        </ul>
    </td>
```

File: crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_checkboxes_false.html (hunk 2 above)
Comment (smithdc1) [reply crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_checkboxes_false.html:5]: See #1156

File: crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_checkboxes_false.html (hunk 3 above)
Comment (smithdc1) [reply crispy_forms/tests/results/bootstrap4/test_layout/test_use_custom_control_is_used_in_checkboxes_false.html:2]: I think this is fine. While Django puts the first inputs reference here the [bootstrap docs](https://getbootstrap.com/docs/4.6/components/forms/#horizontal-form) don't show this and accessibility tests don't complain either.

Wave complains about the layout but that's due to not using a fieldset and a legend. This is a separate issue.

**Comment by ritiko** [comment]
Hi @smithdc1 could you please re-examine this PR. It seems like it's been the cause of quite a number of issues raised for django-crispy-forms. I think we could probably find a way to re-do this that doesn't necessarily break a lot of the existing implementations. I ask because you're a lot more familiar with the code base. All the same, I want to really thank you for everything you've done to make this possible.
